### PR TITLE
Add statefulset.kubernetes.io/pod-ordinal label

### LINKED ISSUE #30427: Consider exposing petset index via downward api
While it's easy enough to get through something like:
`IFS='-' read -ra LIST <<< "$POD_NAME"; INDEX=${LIST[1]}`
It requires a custom entrypoint. If the ordinal index is a common part of 12 fa apps we should just expose it more naturally through the downward api. 

ref https://github.com/kubernetes/kubernetes/issues/30256
@tobegit3hub

### LINKED ISSUE #40651: Are there plans to make the ordinal index for a statefulset available dynamically within a config file?
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): statefulset ordinal config
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Feature Request

**Kubernetes version** (use `kubectl version`): 1.5.2

With a StatefulSet, the hostname is set to `name-`_ordinal_. It is a common pattern with the types of services that go into StatefulSets to expect the unique ID of the service to be passed, e.g. as an env var. 

This causes all sorts of interesting contortions, like extending the official zookeeper image to try and guess the ordinal index of the set from the hostname and then set another var with the index.

It would be good if there were some templating format so that the property could be set directly in the config file. E.g. (simplified):

```yml
apiVersion: apps/v1beta1
kind: StatefulSet
metadata:
  name: zk
spec:
  serviceName: zk-headless
  replicas: 3
  template:
    metadata:
      labels:
        app: zk
    spec:
      containers:
      - name: zk
        imagePullPolicy: Always
        image: zookeeper
        resources:
        env:
        - name : ZOO_MY_ID
          value: ${{ordinal}}
        - name : ZOO_SERVERS
          value: server.0=zk-0:2888:3888 server.1=zk-1:2888:3888 server.2=zk-2:2888:3888
        - name : ZOO_TICK_TIME
          value: "2000"
        - name : ZOO_INIT_LIMIT
          value: "10"
        - name : ZOO_SYNC_LIMIT
          value: "5"
```

### PR DESCRIPTION
**Pull request by jlegrone**
**What this PR does / why we need it**:
This PR makes the ordinal number of each pod in a statefulset available with the label `statefulset.kubernetes.io/pod-ordinal`.

This enables direct access to each pod’s ordinal number as an environment variable via the downward API:

```yaml
env:
- name: POD_ORDINAL
  valueFrom:
    fieldRef:
      fieldPath: metadata.labels['statefulset.kubernetes.io/pod-ordinal']
```

**Which issue(s) this PR fixes**:
Fixes #30427 and fixes #40651 

**Special notes for your reviewer**:

**Release note**:

```release-note
StatefulSet controller will create a label, `statefulset.kubernetes.io/pod-ordinal`, for the ordinal number of each Pod in a StatefulSet. This enables access to the ordinal number of each Pod controlled by a StatefulSet via the downward API.
```

### CODE REVIEW

**Review by dixudx — CHANGES_REQUESTED** [review]
We have `pod.Name` and `pod.GenerateName`, where we can get the ordinal number. I don't think we should have such `statefulset.kubernetes.io/pod-ordinal` label.

**Comment by jlegrone** [comment]
@dixudx I certainly agree that it is trivial to parse the ordinal number from the pod name. It is a usability concern, however, as outlined by many comments in the linked github issues. Directly accessing the ordinal number as an environment variable is less brittle (every user does not need to implement their own logic) and more flexible (no need to override the entrypoint or use pod lifecycle events in order to support "off the shelf" images as communicated in comments on linked issues).

**Comment by enisoc** [comment]
/ok-to-test

This makes sense to me, having written my fair share of annoying entrypoint wrapper scripts to parse the Pod name via `$(hostname)` into an ordinal env var. The addition of the `pod-name` label helped a little, but you still need the ugly wrapper script to parse the Pod name if you want the ordinal, and I've seen many people get this wrong (e.g. their scripts get confused if the StatefulSet name itself also contains a match for `-(\d+)`).

For anyone wondering why we couldn't have just added `pod-ordinal` to begin with instead of `pod-name`, it's because `pod-name` addresses a different use case. The purpose of `pod-name` was to have a label selector `pod-name: blah-N` that's guaranteed to select only one Pod in a given namespace. By contrast, a selector built from (StatefulSet's label selector + `pod-ordinal: N`) is not guaranteed to only select one Pod because multiple StatefulSets' label selectors can overlap or even be identical.

I think it makes sense to have both labels. The ordinal of a StatefulSet Pod meets the criteria to be a label because it's part of the identity of the Pod, and the linked issues show that it's useful.

Does anyone have any remaining objections to this?

**Review by tnozicka — APPROVED** [review]
I like the label and the DownwardAPI. Parsing a pod name is certainly not the interface we should be providing our users to get the ordinal value.

/lgtm

File: pkg/controller/statefulset/stateful_set_utils.go
Comment (smarterclayton) [inline pkg/controller/statefulset/stateful_set_utils.go]: This will break on upgrade - this check should not be added (it's not backwards compatible when viewing older stateful sets).  We shouldn't be checking this anyway because podName is canonical and ordinal is a secondary attribute of name..

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (smarterclayton) [inline staging/src/k8s.io/api/apps/v1/types.go]: Please add good description for this label, and also add godoc to the v1 StatefulSet struct that describes the labels created on each pod in the set.

**Review by smarterclayton — CHANGES_REQUESTED** [review]
You need to add an e2e test for this label.

Also, you need to define ordinal in godoc on v1 Statefulset (what the allowed values are).

Finally, we need to decide whether we are allowed to bypass an alpha phase for the label - other api approvers need to weigh in.

**Comment by smarterclayton** [comment]
/hold

while we resolve the alpha API question @kubernetes/api-reviewers

**Comment by smarterclayton** [comment]
Can you articulate the use case for selecting by an ordinal as justification for this not just being an annotation (which is a smaller API change)?

**Comment by jlegrone** [comment]
> Can you articulate the use case for selecting by an ordinal as justification for this not just being an annotation (which is a smaller API change)?

@smarterclayton I don't think it will ever be a common case, but if a user (or more likely a CRD controller) is orchestrating several statefulsets to work together in some way, for example to represent replicated shards in a database, then it may be desirable to select a cross-section of pods controlled by those statefulsets.

Given this topology:

| sfs `shard-a`             | sfs `shard-b`             | sfs `shard-c`             |
|---------------------------|---------------------------|---------------------------|
| master (pod `shard-a-0`)  | master (pod `shard-b-0`)  | master (pod `shard-c-0`)  |
| replica (pod `shard-a-1`) | replica (pod `shard-b-1`) | replica (pod `shard-c-1`) |
| replica (pod `shard-a-2`) | replica (pod `shard-b-2`) | replica (pod `shard-c-2`) |

Users could leverage a label to view the state of only `replica` pods:

`kubectl get pods -l tier=db,statefulset.kubernetes.io/pod-ordinal!=0`

Or define a service to load balance only to `master` pods:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: db-read-write
spec:
  selector:
    tier: db
    statefulset.kubernetes.io/pod-ordinal: "0"
```

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (jlegrone) [reply staging/src/k8s.io/api/apps/v1/types.go]: Added documentation:

screen shot 2018-09-20 at 3 06 11 pm

Very open to feedback on wording here...

**Comment by jlegrone** [comment]
@smarterclayton would you mind elaborating on what an e2e for this change should look like? I don't see any e2e testing for `statefulset.kubernetes.io/pod-name` at the moment -- I wouldn't mind adding one if you have ideas about the design.

Also, is an e2e test necessary or is an additional package test enough?

File: pkg/controller/statefulset/stateful_set_utils.go
Comment (jlegrone) [reply pkg/controller/statefulset/stateful_set_utils.go]: Are we sure this will break on upgrade? It looks like the controller updates pod identity if there is not a match:

https://github.com/kubernetes/kubernetes/blob/337e0e18f1aefa199bd0a1786f8eab42e948064c/pkg/controller/statefulset/stateful_pod_control.go#L95-L98

Since I implemented this in the `updateIdentity` function it seemed to make sense to also update `identityMatches`. I'll wait until we can get more feedback on label vs annotation to change this.

**Comment by jlegrone** [comment]
@smarterclayton any updates?

**Comment by smarterclayton** [comment]
The e2e test would verify that the label and ordinal is set, that none overlap, and that during replacement it is not inaccurate.  Adding it to the existing e2e stateful set tests is sufficient.

**Comment by jberkus** [comment]
Is this PR held up by anything other than the rebase and the need for e2e tests, right now?

**Comment by smarterclayton** [comment]
@liggitt do you have an opinion on bypassing alpha for this label?  I think I'd say we follow field rules here.   We *could* put the label behind a feature gate for one release, which also accomplishes this goal.

**Comment by jlegrone** [comment]
hmm can't reproduce test error locally

/retest

**Comment by liggitt** [comment]
> @liggitt do you have an opinion on bypassing alpha for this label? I think I'd say we follow field rules here. We _could_ put the label behind a feature gate for one release, which also accomplishes this goal.

I assume you mean whether to add the label at all, rather than whether to include ".alpha." in the label key?

Because the fields are in metadata, there's no loss of data in skew or downgrade scenarios.

On upgrade, will the controller add the label to existing pods?

File: pkg/controller/statefulset/stateful_set_utils.go
Comment (jlegrone) [reply pkg/controller/statefulset/stateful_set_utils.go]: @liggitt 
> On upgrade, will the controller add the label to existing pods?

I believe so; see above comment.

**Comment by jlegrone** [comment]
@liggitt 
> On upgrade, will the controller add the label to existing pods?

I believe so; see this comment: https://github.com/kubernetes/kubernetes/pull/68719/files#r219337216

**Comment by jlegrone** [comment]
@liggitt do you mind taking a look at the test cases that have been added?

**Comment by so0k** [comment]
Any progress on getting this merged?

**Comment by kow3ns** [comment]
The name label is to make each Pod selectable by name for the purposes of attaching a service to it. It wasn't for downward API injection. You can already get the name via the [metadata API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/). If the only purpose of this is to inject the ordinal via the downward-api, I think an annotation is better suited and conforms to convention.

**Comment by jlegrone** [comment]
> If the only purpose of this is to inject the ordinal via the downward-api, I think an annotation is better suited and conforms to convention.

@kow3ns this is not the case. Please see additional use cases outlined in https://github.com/kubernetes/kubernetes/pull/68719#issuecomment-423259874, which would not be possible if ordinal were only available as an annotation.

**Comment by jfrabaute** [comment]
I sent a similar PR for this (I'd be happy to abandon it if this one is merged).

But for some application, "0" is not allowed as an ID (zookeeper for instance), so I've added 2 annotations, (id0/id1), similar to some template engines exposing two indices for loops in case 0 or 1 at first id is needed.
Would it be possible to add something similar on this PR?

PR #83101

**Comment by alejandrox1** [comment]
Thank you for working on this @jlegrone !
What do we need (api wise) to get this in @liggitt ?

/remove-area e2e-test-framework
/milestone v1.17
/ok-to-test

**Comment by liggitt** [comment]
> What do we need (api wise) to get this in @liggitt ?

Added to the API review tracking board. It looks like @smarterclayton is the apps api approver most involved with the discussion on this. Reach out to him to schedule a time for the API review.

It would also be good to resolve https://github.com/kubernetes/kubernetes/pull/68719#issuecomment-534889527 and the competing implementation that provides alternate indexing (https://github.com/kubernetes/kubernetes/pull/83101). My first thought is that adaptation of indexes to a different starting point could be done in the container, rather than providing multiple indexing labels on the pod.

**Comment by jfrabaute** [comment]
> It would also be good to resolve [#68719 (comment)](https://github.com/kubernetes/kubernetes/pull/68719#issuecomment-534889527) and the competing implementation that provides alternate indexing (#83101). 

Thanks for taking the time to look at this other option.

> My first thought is that adaptation of indexes to a different starting point could be done in the container, rather than providing multiple indexing labels on the pod.

I agree in general, but I've added "1" as it's a "special" case that might be pretty common.
Right now, here is what I have for zookeeper (initContainer):

        command:
        - bash
        - "-c"
        - "[ -e /zookeeper/data/myid ] || echo $((${PODNAME##*-}+1)) > /zookeeper/data/myid"

So, having the new annotation starting at 0 will not change the fact that I still need to do the +1 from the value I get from the env var. And then with the DownwardAPIVolumeFiles, I can directly mount this value as a file (as I have the index starting a 1) and I don't even need this initContainer.

Having said that, just getting the index starting at zero is already a good improvement.

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (smarterclayton) [inline staging/src/k8s.io/api/apps/v1/types.go]: Can pod name be longer than the allowed label length?

Comment (smarterclayton) [reply staging/src/k8s.io/api/apps/v1/types.go]: Why do we need this label when you can already get pod name in downward API?

**Review by smarterclayton — COMMENTED** [review]
A few questions.

**Comment by ttousai** [comment]
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on Nov. 18. Is this PR still intended for 1.17?

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (jlegrone) [reply staging/src/k8s.io/api/apps/v1/types.go]: >Why do we need this label when you can already get pod name in downward API?

I believe `statefulset.kubernetes.io/pod-name` is applied in order to support creating a unique service per pod. But please note that this changeset doesn't add the pod-name label; I only added a comment for the existing `StatefulSetPodNameLabel` const.

> Can pod name be longer than the allowed label length?

Yes, and this is a common (existing) problem with statefulsets. See https://github.com/kubernetes/kubernetes/issues/64023.

**Comment by ttousai** [comment]
***Correction:*** Code freeze is on Thursday, November 14.

Sorry for the mix up.

**Comment by ttousai** [comment]
@jlegrone  will this PR be ready for code freeze this Thursday 14th November?

**Comment by annajung** [comment]
Hello @jlegrone  👋 I'm one of the Bug Triage shadows for the 1.18 release. Since this PR is tagged for 1.18, I'd like to check its status. Is this PR intended for milestone 1.18? 

Friendly reminder: Code freeze is scheduled for March 5th.

**Comment by jlegrone** [comment]
Thanks @annajung, according to https://github.com/kubernetes/kubernetes/pull/68719#issuecomment-538489175 I think we're currently awaiting further review from @smarterclayton or another representative of sig apps. I doubt that this is high priority to get into the 1.18 release, but I would certainly love to either merge or close this PR so I can finally remove it from my dashboard! 😄

**Comment by annajung** [comment]
Hey @jlegrone Thanks for getting back to me! I will keep this PR tracked in the bug triage tracking sheet for release 1.18 for now, hopefully, it gets reviewed soon to make it in the upcoming release!

**Comment by annajung** [comment]
Hey @jlegrone @smarterclayton Checking in again. Do you think this will make it in the 1.18 release? 
Friendly reminder: Code freeze is scheduled for March 5th.

File: test/e2e/framework/statefulset/rest.go
Comment (tedyu) [inline test/e2e/framework/statefulset/rest.go]: Is this check necessary (considering the name check on line 284) ?

**Comment by smarterclayton** [comment]
Putting this back on my priority list

**Comment by annajung** [comment]
Hey @smarterclayton Checking in again as code freeze is coming up in a few days. Do you think this will make it into the 1.18 release?

**Comment by annajung** [comment]
Just a friendly reminder that code freeze is in 2 days @jlegrone @smarterclayton

**Comment by robermar23** [comment]
Would love to see this addition merged and released.  Would help with assignment of graylog public_uri variable that needs to be unique per node in a graylog statefulset

**Comment by droslean** [comment]
Moving this to 1.20. 

/milestone v1.20

**Comment by duglin** [comment]
@dims is that wishful thinking or @liggitt are you gonna help nudge this one over the finish line?

File: test/e2e/framework/statefulset/rest.go
Comment (mikebrow) [reply test/e2e/framework/statefulset/rest.go:290]: looks right as this is the non overlapping check

**Comment by Huang-Wei** [comment]
It seems most people here think it's a useful feature, is there any way we can move it forward?

**Comment by dieucao** [comment]
I was curious what the status of this PR is? 
I see that @liggitt moved this to "Changes Requested" in API Reviews a couple weeks ago but I wasn't able to tell if there are any recent comments/change requests for what is needed at this point from @jlegrone to move this PR forward?

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (smarterclayton) [reply staging/src/k8s.io/api/apps/v1/types.go:33]: Ok, let's remove all pod name references in here and the docs then.

File: staging/src/k8s.io/api/apps/v1/generated.proto
Comment (smarterclayton) [inline staging/src/k8s.io/api/apps/v1/generated.proto:536]: `pods` is appropriate, not `Pods`.

File: test/e2e/framework/statefulset/rest.go
Comment (smarterclayton) [inline test/e2e/framework/statefulset/rest.go:280]: Remove this since we should no longer be setting it.

File: staging/src/k8s.io/api/apps/v1/generated.proto
Comment (smarterclayton) [inline staging/src/k8s.io/api/apps/v1/generated.proto:549]: Document the behavior if a user sets this label on the statefulset pod template.  What do we do if the user sets the replica set label?

Comment (smarterclayton) [reply staging/src/k8s.io/api/apps/v1/generated.proto:549]: (document and/or determine whether we need to add validation logic)

**Comment by bai** [comment]
Hey 👋 Bug Triage here. This PR has not been updated for a while, so I'd like to check on the status. The code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.20, is it still planned for this release?

**Comment by AWKIF** [comment]
would be great if this happen for 1.20 :+1:

**Comment by cuttingedge1109** [comment]
Seems need to rebase before merge

**Comment by desponda** [comment]
Hello! I am from the Bug Triage team! This PR has not been updated for a long time, so I'd like to check on the status @jlegrone . The code freeze is starting March 9th, 2021 (about 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze.
As the PR|Issue is tagged for 1.21, is it still planned for this release?

**Comment by shanefay422** [comment]
Is this PR going to be pushed over the line anytime in the foreseeable future?

**Comment by jlegrone** [comment]
👋 I've let this PR sit for far too long. It seems like there was consensus at one point that this was an acceptable change, but I don't have the bandwidth to see this through at the moment.

I'm closing now to make it clear that this feature is not in progress and should be picked up by someone else with the interest and time. Thank you to the reviewers, and the release team members who have been checking in for the last 3 years! I hope I didn't waste too much of your time. 😅

**Comment by jaigopinathmalempati** [comment]
> **What this PR does / why we need it**: This PR makes the ordinal number of each pod in a statefulset available with the label `statefulset.kubernetes.io/pod-ordinal`.
> 
> This enables direct access to each pod’s ordinal number as an environment variable via the downward API:
> 
> ```yaml
> env:
> - name: POD_ORDINAL
>   valueFrom:
>     fieldRef:
>       fieldPath: metadata.labels['statefulset.kubernetes.io/pod-ordinal']
> ```
> 
> **Which issue(s) this PR fixes**: Fixes #30427 and fixes #40651
> 
> **Special notes for your reviewer**:
> 
> **Release note**:
> 
> ```
> StatefulSet controller will create a label, `statefulset.kubernetes.io/pod-ordinal`, for the ordinal number of each Pod in a StatefulSet. This enables access to the ordinal number of each Pod controlled by a StatefulSet via the downward API.
> ```

tried using it but no luck . Can someone help here
My current K8s version is:
Client Version: v1.27.2
Kustomize Version: v5.0.1
Server Version: v1.26.5

**Pull request closed without merging**
